bindings/cpp/src/ethereum.rs

```diff
@@ -0,0 +1,306 @@
+            let token = EthAbiToken::Array(tokens);
+            self.state = EthContractState::Value;
+            self.clear_tokens()?;
+            Ok(Box::new(EthAbiTokenWrapper { token }))
+        } else {
+            bail!("tmptokens is None")
+        }
+    }
+
+    // tuple
+    fn begin_tuple(&mut self) -> Result<()> {
+        if self.tmptokens.is_none() {
+            self.tmptokens = Some(Vec::new());
+            self.state = EthContractState::Tuple;
+        } else {
+            bail!("tmptokens is not None")
+        }
+        Ok(())
+    }
+    fn commit_tuple(&mut self) -> Result<Box<EthAbiTokenWrapper>> {
+        if self.state != EthContractState::Tuple {
+            bail!("state is not Tuple")
+        }
+        if let Some(tmptokens) = self.tmptokens.take() {
+            let tokens = tmptokens;
+            let token = EthAbiToken::Tuple(tokens);
+            self.clear_tokens()?;
+            Ok(Box::new(EthAbiTokenWrapper { token }))
+        } else {
+            bail!("tmptokens is None")
+        }
+    }
+
+    pub fn decode_input(&mut self, function_name: &str, data: &[u8]) -> Result<String> {
+        let jsonoutput = self.abi_contract.decode_input(function_name, data)?;
+        Ok(jsonoutput)
+    }
+
+    pub fn decode_output(&mut self, function_name: &str, data: &[u8]) -> Result<String> {
+        let jsonoutput = self.abi_contract.decode_output(function_name, data)?;
+        Ok(jsonoutput)
+    }
+
+    pub fn encode(&mut self, function_name: &str) -> Result<Vec<u8>> {
+        let tokens = self.tokens.clone();
+        let srcbytes = self.abi_contract.encode(function_name, tokens)?;
+        Ok(srcbytes)
+    }
+
+    fn clear_tokens(&mut self) -> Result<()> {
+        self.tokens.clear();
+        self.tmptokens = None;
+        self.state = EthContractState::Value;
+        Ok(())
+    }
+
+    pub async fn do_call(
+        &mut self,
+        rpcserver: &str,
+        contract_address: &str,
+        function_name: &str,
+    ) -> Result<String> {
+        let addr = Address::from_str(contract_address)?;
+        let provider = Provider::<Http>::try_from(rpcserver)?;
+        let data = self.tokens.clone();
+        let srcbytes = self.abi_contract.encode(function_name, data)?;
+        let ethbytes = Bytes::from(srcbytes);
+        let tx = Eip1559TransactionRequest::new().data(ethbytes).to(addr);
+        let tx = TypedTransaction::Eip1559(tx);
+        let response: Bytes = provider.call(&tx, None).await?;
+        let jsondata = self.decode_output(function_name, &response)?;
+        Ok(jsondata)
+    }
+    pub fn call(
+        &mut self,
+        rpcserver: &str,
+        contract_address: &str,
+        function_name: &str,
+    ) -> Result<String> {
+        let rt = tokio::runtime::Runtime::new().map_err(|_err| EthError::AsyncRuntimeError)?;
+        let res = rt.block_on(self.do_call(rpcserver, contract_address, function_name))?;
+        Ok(res)
+    }
```

Comment:
you can use `DynamicContract` as mentioned on Slack:
1. `let dc = DynamicContract::new(…)`
2. `let contract_call = dc.function_call(…)`

Reply:
ok, i'll modify to use DynamicContract